### ISSUE TITLE
gunicorn no response intermittently issue after max_requests restart

### ISSUE THREAD

**Issue by inyong7ee** [benoitc/gunicorn#2021]
Hello,

This issue occasionally occurs like once every ten days. 
We run two production servers using gunicorn 19.9.0 with Django for sync job and rabbitmq + celery for async task.

With gunicorn setting like below:
workers = cpu_count()+1 # which is 17
max_requests = 1000
max_requests_jitter = 30
worker_class = 'gevent'
timeout = 30
loglevel = "info"

Server: 16 cores, 128GB ram, CentOS 7.5.1804
Restart server everyday

When one gunicorn worker restarts after max_requests API calls, some API calls don't respond at all so the host timeout after 2 min. These calls don't leave any gunicorn errorlog messages like TIMEOUT. 
After one gunicorn worker restarted, most API calls works fine and a few non-response calls happen intermittently and continuosly. 

APIs are only called locally, by Spring framework in the same instance, and non-response issues do not occur only in certain APIs.
These are mainly for data analytics service and the response time is varying for each API, which is about 100 ~ 1300ms.

I would be grateful if you could give me some advice to resolve this issue.

#errorlog:

> [2019-04-28 06:07:38 +0900] [122099] [INFO] Starting gunicorn 19.9.0
[2019-04-28 06:07:38 +0900] [122099] [ERROR] Connection in use: ('0.0.0.0', 8090)
[2019-04-28 06:07:38 +0900] [122099] [ERROR] Retrying in 1 second.
[2019-04-28 06:07:38 +0900] [39324] [INFO] Worker exiting (pid: 39324)
[2019-04-28 06:07:38 +0900] [39303] [INFO] Worker exiting (pid: 39303)
[2019-04-28 06:07:38 +0900] [39308] [INFO] Worker exiting (pid: 39308)
[2019-04-28 06:07:38 +0900] [39307] [INFO] Worker exiting (pid: 39307)
[2019-04-28 06:07:38 +0900] [39311] [INFO] Worker exiting (pid: 39311)
[2019-04-28 06:07:38 +0900] [39302] [INFO] Worker exiting (pid: 39302)
[2019-04-28 06:07:38 +0900] [39329] [INFO] Worker exiting (pid: 39329)
[2019-04-28 06:07:38 +0900] [39312] [INFO] Worker exiting (pid: 39312)
[2019-04-28 06:07:38 +0900] [39310] [INFO] Worker exiting (pid: 39310)
[2019-04-28 06:07:38 +0900] [109617] [INFO] Worker exiting (pid: 109617)
[2019-04-28 06:07:38 +0900] [39317] [INFO] Worker exiting (pid: 39317)
[2019-04-28 06:07:38 +0900] [120397] [INFO] Worker exiting (pid: 120397)
[2019-04-28 06:07:38 +0900] [39322] [INFO] Worker exiting (pid: 39322)
[2019-04-28 06:07:38 +0900] [39313] [INFO] Worker exiting (pid: 39313)
[2019-04-28 06:07:38 +0900] [39332] [INFO] Worker exiting (pid: 39332)
[2019-04-28 06:07:38 +0900] [104426] [INFO] Worker exiting (pid: 104426)
[2019-04-28 06:07:38 +0900] [39318] [INFO] Worker exiting (pid: 39318)
[2019-04-28 06:07:39 +0900] [39299] [INFO] Shutting down: Master
[2019-04-28 06:07:39 +0900] [122099] [INFO] Listening at: http://0.0.0.0:8090 (122099)
[2019-04-28 06:07:39 +0900] [122099] [INFO] Using worker: gevent
[2019-04-28 06:07:39 +0900] [122104] [INFO] Booting worker with pid: 122104
[2019-04-28 06:07:39 +0900] [122105] [INFO] Booting worker with pid: 122105
[2019-04-28 06:07:39 +0900] [122107] [INFO] Booting worker with pid: 122107
[2019-04-28 06:07:39 +0900] [122109] [INFO] Booting worker with pid: 122109
[2019-04-28 06:07:39 +0900] [122111] [INFO] Booting worker with pid: 122111
[2019-04-28 06:07:39 +0900] [122113] [INFO] Booting worker with pid: 122113
[2019-04-28 06:07:39 +0900] [122115] [INFO] Booting worker with pid: 122115
[2019-04-28 06:07:39 +0900] [122120] [INFO] Booting worker with pid: 122120
[2019-04-28 06:07:40 +0900] [122122] [INFO] Booting worker with pid: 122122
[2019-04-28 06:07:40 +0900] [122125] [INFO] Booting worker with pid: 122125
[2019-04-28 06:07:40 +0900] [122126] [INFO] Booting worker with pid: 122126
[2019-04-28 06:07:40 +0900] [122128] [INFO] Booting worker with pid: 122128
[2019-04-28 06:07:40 +0900] [122133] [INFO] Booting worker with pid: 122133
[2019-04-28 06:07:40 +0900] [122135] [INFO] Booting worker with pid: 122135
[2019-04-28 06:07:40 +0900] [122138] [INFO] Booting worker with pid: 122138
[2019-04-28 06:07:40 +0900] [122140] [INFO] Booting worker with pid: 122140
[2019-04-28 06:07:40 +0900] [122144] [INFO] Booting worker with pid: 122144
[2019-04-28 11:07:57 +0900] [122144] [INFO] Autorestarting worker after current request.
[2019-04-28 11:07:59 +0900] [122144] [INFO] Worker exiting (pid: 122144)
[2019-04-28 11:07:59 +0900] [56720] [INFO] Booting worker with pid: 56720
>
>**(Here it happens)**
>
>[2019-04-28 12:04:44 +0900] [122099] [INFO] Handling signal: term
[2019-04-28 12:04:44 +0900] [122126] [INFO] Worker exiting (pid: 122126)
[2019-04-28 12:04:44 +0900] [122125] [INFO] Worker exiting (pid: 122125)
[2019-04-28 12:04:44 +0900] [122104] [INFO] Worker exiting (pid: 122104)
[2019-04-28 12:04:44 +0900] [122105] [INFO] Worker exiting (pid: 122105)
[2019-04-28 12:04:44 +0900] [122107] [INFO] Worker exiting (pid: 122107)
[2019-04-28 12:04:44 +0900] [122133] [INFO] Worker exiting (pid: 122133)
[2019-04-28 12:04:44 +0900] [122109] [INFO] Worker exiting (pid: 122109)
[2019-04-28 12:04:44 +0900] [122128] [INFO] Worker exiting (pid: 122128)
[2019-04-28 12:04:44 +0900] [122122] [INFO] Worker exiting (pid: 122122)
[2019-04-28 12:04:44 +0900] [122138] [INFO] Worker exiting (pid: 122138)
[2019-04-28 12:04:44 +0900] [122140] [INFO] Worker exiting (pid: 122140)
[2019-04-28 12:04:44 +0900] [122111] [INFO] Worker exiting (pid: 122111)
[2019-04-28 12:04:44 +0900] [122135] [INFO] Worker exiting (pid: 122135)
[2019-04-28 12:05:14 +0900] [122099] [INFO] Shutting down: Master
[2019-04-28 12:05:55 +0900] [70005] [INFO] Starting gunicorn 19.9.0
[2019-04-28 12:05:55 +0900] [70005] [INFO] Listening at: http://0.0.0.0:8090 (70005)
[2019-04-28 12:05:55 +0900] [70005] [INFO] Using worker: gevent
[2019-04-28 12:05:55 +0900] [70008] [INFO] Booting worker with pid: 70008
[2019-04-28 12:05:55 +0900] [70009] [INFO] Booting worker with pid: 70009
[2019-04-28 12:05:55 +0900] [70011] [INFO] Booting worker with pid: 70011
[2019-04-28 12:05:55 +0900] [70013] [INFO] Booting worker with pid: 70013
[2019-04-28 12:05:55 +0900] [70015] [INFO] Booting worker with pid: 70015
[2019-04-28 12:05:55 +0900] [70016] [INFO] Booting worker with pid: 70016
[2019-04-28 12:05:55 +0900] [70020] [INFO] Booting worker with pid: 70020
[2019-04-28 12:05:55 +0900] [70022] [INFO] Booting worker with pid: 70022
[2019-04-28 12:05:56 +0900] [70026] [INFO] Booting worker with pid: 70026
[2019-04-28 12:05:56 +0900] [70030] [INFO] Booting worker with pid: 70030
[2019-04-28 12:05:56 +0900] [70033] [INFO] Booting worker with pid: 70033
[2019-04-28 12:05:56 +0900] [70036] [INFO] Booting worker with pid: 70036
[2019-04-28 12:05:56 +0900] [70039] [INFO] Booting worker with pid: 70039
[2019-04-28 12:05:56 +0900] [70042] [INFO] Booting worker with pid: 70042
[2019-04-28 12:05:56 +0900] [70044] [INFO] Booting worker with pid: 70044
[2019-04-28 12:05:56 +0900] [70047] [INFO] Booting worker with pid: 70047
[2019-04-28 12:05:56 +0900] [70051] [INFO] Booting worker with pid: 70051

Thank you.

**Comment by inyong7ee** [comment]
I've set logs to server hook funtions in the config src file.

    # Server Hooks for debugging

    def post_fork(server, worker):
        # Called just after a worker has been forked.
        worker.log.debug("[post_fork] worker spawned - pid %s" % worker.pid)
        patch_psycopg()

    def on_starting(server):
        # Called just before the master process is initialized.
        server.log.debug("[on_starting] Arbiter")

    def on_reload(server):
        # Called to recycle workers during a reload via SIGHUP.
        server.log.debug("[on_reload] Arbiter")

    def when_ready(server):
        # Called just after the server is started.
        server.log.debug("[when_ready] Arbiter - Server is ready. Spawning workers")

    def pre_fork(server, worker):
        # Called just before a worker is forked.
        worker.log.debug("[pre_fork] worker pid %s" % worker.pid)

    def post_worker_init(worker):
        # Called just after a worker has initialized the application.
        worker.log.debug("[post_worker_init] worker pid %s" % worker.pid)

    def worker_int(worker):
        # Called just after a worker exited on SIGINT or SIGQUIT.
        worker.log.debug("[worker_int] worker received INT or QUIT signal - pid %s" % worker.pid)

        ## get traceback info
        import threading, sys, traceback
        id2name = {th.ident: th.name for th in threading.enumerate()}
        code = []
        for threadId, stack in sys._current_frames().items():
            code.append("\n# Thread: %s(%d)" % (id2name.get(threadId,""),
                threadId))
            for filename, lineno, name, line in traceback.extract_stack(stack):
                code.append('File: "%s", line %d, in %s' % (filename, lineno, name))
                if line:
                    code.append("[worker_int] %s" % (line.strip()))
        worker.log.debug("\n".join(code))

    def worker_abort(worker):
        # Called when a worker received the SIGABRT signal. generally happens on timeout.
        worker.log.debug("[worker_abort] worker received SIGABRT signal - pid %s" % worker.pid)

    def pre_exec(server):
        # Called just before a new master process is forked.
        server.log.debug("[pre_exec] Arbiter - Forked child, re-executing.")

    def pre_request(worker, req):
        # Called just before a worker processes the request.
        worker.log.debug("[pre_request] worker pid %s | %s | %s" % (worker.pid, req.method, req.path))

    def post_request(worker, req, environ, resp):
        # Called after a worker processes the request.
        worker.log.debug("[post_request] worker pid %s | %s | %s" % (worker.pid, req.method, req.path))

    def child_exit(server, worker):
        # Called just after a worker has been exited, in the master process.
        server.log.debug("[child_exit] worker pid %s" % worker.pid)

    def worker_exit(server, worker):
        # Called just after a worker has been exited, in the worker process.
        server.log.debug("[worker_exit] worker pid %s" % worker.pid)

    def nworkers_changed(server, new_value, old_value):
        # Called just after num_workers has been changed.
        server.log.debug("[nworkers_changed] Arbiter")

    def on_exit(server):
        # Called just before exiting Gunicorn.
        server.log.debug("[on_exit] Arbiter")

        
After 11 days since the last error, I got the log.
The process 96122 is the Arbiter and the worker 73015 is the problem. 

> [2019-05-09 15:03:40 +0900] [96166] [DEBUG] [pre_request] worker pid 96166 | POST | /analyapi/kt_drl_lssn_achv_info
> **[2019-05-09 15:03:40 +0900] [96166] [INFO] Autorestarting worker after current request.**
> [2019-05-09 15:03:40 +0900] [96166] [DEBUG] [post_request] worker pid 96166 | POST | /analyapi/kt_drl_lssn_achv_info
> [2019-05-09 15:03:40 +0900] [96166] [DEBUG] Closing connection. 
> **[2019-05-09 15:03:41 +0900] [96166] [INFO] Worker exiting (pid: 96166)**
> [2019-05-09 15:03:41 +0900] [96166] [DEBUG] [worker_exit] worker pid 96166
> [2019-05-09 15:03:41 +0900] [96122] [DEBUG] [child_exit] worker pid 96166
> [2019-05-09 15:03:42 +0900] [96122] [DEBUG] [pre_fork] worker pid [booting]
> **[2019-05-09 15:03:42 +0900] [73015] [INFO] Booting worker with pid: 73015**
> **[2019-05-09 15:03:42 +0900] [73015] [DEBUG] [post_fork] worker spawned - pid 73015**
> **[2019-05-09 15:03:42 +0900] [73015] [DEBUG] [post_worker_init] worker pid 73015**
> [2019-05-09 15:03:44 +0900] [96160] [DEBUG] [pre_request] worker pid 96160 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:03:44 +0900] [96160] [DEBUG] [post_request] worker pid 96160 | POST | /analyapi/sr_rcmnd_recall_qst
> **[2019-05-09 15:03:46 +0900] [73015] [DEBUG] [pre_request] worker pid 73015 | POST | /analyapi/lp_prdt_tlrn_achv**
> [2019-05-09 15:03:46 +0900] [96158] [DEBUG] [pre_request] worker pid 96158 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:03:46 +0900] [96158] [DEBUG] [post_request] worker pid 96158 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:03:47 +0900] [96158] [DEBUG] [pre_request] worker pid 96158 | POST | /analyapi/kt_drl_lssn_achv_info
> [2019-05-09 15:03:48 +0900] [96158] [DEBUG] [post_request] worker pid 96158 | POST | /analyapi/kt_drl_lssn_achv_info
> [2019-05-09 15:03:49 +0900] [96158] [DEBUG] Closing connection. 
> [2019-05-09 15:03:54 +0900] [96158] [DEBUG] [pre_request] worker pid 96158 | POST | /analyapi/ir_weak_stdy_drl_qst
> [2019-05-09 15:03:54 +0900] [96158] [DEBUG] [post_request] worker pid 96158 | POST | /analyapi/ir_weak_stdy_drl_qst
> [2019-05-09 15:03:56 +0900] [96158] [DEBUG] [pre_request] worker pid 96158 | POST | /analyapi/ir_weak_stdy_tlrn_qst
> [2019-05-09 15:03:56 +0900] [96158] [DEBUG] [post_request] worker pid 96158 | POST | /analyapi/ir_weak_stdy_tlrn_qst
> [2019-05-09 15:03:57 +0900] [96158] [DEBUG] [pre_request] worker pid 96158 | POST | /analyapi/kt_drl_lssn_achv_info
> [2019-05-09 15:03:58 +0900] [96158] [DEBUG] [post_request] worker pid 96158 | POST | /analyapi/kt_drl_lssn_achv_info
> **[2019-05-09 15:04:04 +0900] [73015] [DEBUG] [pre_request] worker pid 73015 | POST | /analyapi/sr_rcmnd_recall_qst**
> [2019-05-09 15:04:05 +0900] [96160] [DEBUG] [pre_request] worker pid 96160 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:04:06 +0900] [96160] [DEBUG] [post_request] worker pid 96160 | POST | /analyapi/sr_rcmnd_recall_qst
> **[2019-05-09 15:04:16 +0900] [73015] [DEBUG] [pre_request] worker pid 73015 | POST | /analyapi/lp_prdt_tlrn_achv**
> **[2019-05-09 15:04:16 +0900] [73015] [DEBUG] [pre_request] worker pid 73015 | POST | /analyapi/lp_lssn_achv_info**
> [2019-05-09 15:04:19 +0900] [96160] [DEBUG] Closing connection. 
> [2019-05-09 15:04:22 +0900] [96156] [DEBUG] [pre_request] worker pid 96156 | POST | /analyapi/kt_drl_lssn_achv_info
> [2019-05-09 15:04:23 +0900] [96156] [DEBUG] [post_request] worker pid 96156 | POST | /analyapi/kt_drl_lssn_achv_info
> [2019-05-09 15:04:24 +0900] [96156] [DEBUG] [pre_request] worker pid 96156 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:04:24 +0900] [96156] [DEBUG] [post_request] worker pid 96156 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:04:25 +0900] [96156] [DEBUG] [pre_request] worker pid 96156 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:04:26 +0900] [96156] [DEBUG] [post_request] worker pid 96156 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:04:27 +0900] [96156] [DEBUG] [pre_request] worker pid 96156 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:04:28 +0900] [96156] [DEBUG] [post_request] worker pid 96156 | POST | /analyapi/sr_rcmnd_recall_qst
> [2019-05-09 15:04:30 +0900] [96160] [DEBUG] [pre_request] worker pid 96160 | POST | /analyapi/sr_rcmnd_recall_qst
> **[2019-05-09 15:04:30 +0900] [73015] [DEBUG] [pre_request] worker pid 73015 | POST | /analyapi/sr_rcmnd_recall_qst**
> [2019-05-09 15:04:30 +0900] [96160] [DEBUG] [post_request] worker pid 96160 | POST | /analyapi/sr_rcmnd_recall_qst
> **[2019-05-09 15:04:33 +0900] [73015] [DEBUG] [pre_request] worker pid 73015 | POST | /analyapi/sr_rcmnd_recall_qst**
> [2019-05-09 15:04:36 +0900] [96158] [DEBUG] [pre_request] worker pid 96158 | POST | /analyapi/lp_lssn_achv_info
> [2019-05-09 15:04:36 +0900] [96158] [DEBUG] [post_request] worker pid 96158 | POST | /analyapi/lp_lssn_achv_info


The newly autorestarted 73015 wokrer turned into zombie process after getting its first request.

    $ ps -ef | grep defunct

    analy    73075   73015   0 15:03  ?     00:00:00  [gcc] <defunct>

After that, it only gets requests and never processes them.

**Comment by tilgovi** [comment]
Thank you for the detailed logs. I don't know what to think yet, but please report any more information you find.

**Comment by inyong7ee** [comment]
@tilgovi Thank you for the response.

As far as I can tell, this issue happens per every 11 days.
Even if we restart gunicorn and celery everyday morning, it happens on 11th day. It's really strange.

On 11th day, after the first gunicorn worker received max_requests requests, the Arbiter closes it and starts a new gunicorn worker(gevent) and this causes the problem. The newly auto-restarted gunicorn worker doesn't process the first request and makes a child zombie process "[gcc] <defunct>". Afterwards the Arbiter keeps allocating requests to this gunicorn worker and it causes read timeout.

I now doubt the use of sockets or the number of open files in the Linux server.

**Comment by inyong7ee** [comment]
My gunicorn setting:

> [2019-05-10 06:44:25 +0900] [10113] [DEBUG] Current configuration:
>   config: /home/analy/analysystem/analysystem/gunicorn_config.py
>   bind: ['0.0.0.0:8090']
>   backlog: 2048
>   workers: 17
>   worker_class: gevent
>   threads: 1
>   worker_connections: 1000
>   max_requests: 1000
>   max_requests_jitter: 30
>   timeout: 110
>   graceful_timeout: 60
>   keepalive: 2
>   limit_request_line: 4094
>   limit_request_fields: 100
>   limit_request_field_size: 8190
>   reload: False
>   reload_engine: auto
>   reload_extra_files: []
>   spew: False
>   check_config: False
>   preload_app: False
>   sendfile: None
>   reuse_port: False
>   chdir: /home/analy/analysystem
>   daemon: False
>   raw_env: []
>   pidfile: None
>   worker_tmp_dir: None
>   user: 530
>   group: 520
>   umask: 0
>   initgroups: False
>   tmp_upload_dir: None
>   secure_scheme_headers: {'X-FORWARDED-PROTOCOL': 'ssl', 'X-FORWARDED-PROTO': 'https', 'X-FORWARDED-SSL': 'on'}
>   forwarded_allow_ips: ['127.0.0.1']
>   accesslog: /home/logs/analy/gunicorn-access.log
>   disable_redirect_access_to_syslog: False
>   access_log_format: %(h)s %(l)s %(u)s %(t)s "%(r)s" %(s)s %(b)s "%(f)s" "%(a)s" "%(q)s" %(p)s %(D)s
>   errorlog: /home/logs/analy/gunicorn-error.log
>   loglevel: debug
>   capture_output: False
>   logger_class: gunicorn.glogging.Logger
>   logconfig: None
>   logconfig_dict: {}
>   syslog_addr: udp://localhost:514
>   syslog: False
>   syslog_prefix: None
>   syslog_facility: user
>   enable_stdio_inheritance: False
>   statsd_host: None
>   statsd_prefix: 
>   proc_name: None
>   default_proc_name: analysystem.wsgi
>   pythonpath: None
>   paste: None
>   on_starting: <function on_starting at 0x7f5eb35c2620>
>   on_reload: <function on_reload at 0x7f5eb35c26a8>
>   when_ready: <function when_ready at 0x7f5eb35c2730>
>   pre_fork: <function pre_fork at 0x7f5eb35c27b8>
>   post_fork: <function post_fork at 0x7f5eb35c2598>
>   post_worker_init: <function post_worker_init at 0x7f5eb35c2840>
>   worker_int: <function worker_int at 0x7f5eb35c28c8>
>   worker_abort: <function worker_abort at 0x7f5eb35c2950>
>   pre_exec: <function pre_exec at 0x7f5eb35c29d8>
>   pre_request: <function pre_request at 0x7f5eb35c2a60>
>   post_request: <function post_request at 0x7f5eb35c2ae8>
>   child_exit: <function child_exit at 0x7f5eb35c2b70>
>   worker_exit: <function worker_exit at 0x7f5eb35c2bf8>
>   nworkers_changed: <function nworkers_changed at 0x7f5eb35c2c80>
>   on_exit: <function on_exit at 0x7f5eb35c2d08>
>   proxy_protocol: False
>   proxy_allow_ips: ['127.0.0.1']
>   keyfile: None
>   certfile: None
>   ssl_version: 2
>   cert_reqs: 0
>   ca_certs: None
>   suppress_ragged_eofs: True
>   do_handshake_on_connect: False
>   ciphers: TLSv1
>   raw_paste_global_conf: []

**Comment by inyong7ee** [comment]
In order to kill the zombie process, I added some workaround code to server hook functions.
Could you check the os.kill function whether this is right way to kill the not responding worker?


    def post_worker_init(worker):
        # Called just after a worker has initialized the application.
        worker.log.debug("[post_worker_init] worker pid %s" % worker.pid)
        worker.nr_pre = 0
        worker.nr_post = 0

    def pre_request(worker, req):
        # Called just before a worker processes the request.
        worker.log.debug("[pre_request] %s %s - worker pid %s|ppid %s|age %s|nr %s|max_requests %s|alive %s|nr_pre %s|nr_post %s" % (
            req.method, req.path, worker.pid, worker.ppid, worker.age, worker.nr,
            worker.max_requests, worker.alive, worker.nr_pre, worker.nr_post))
        
        if (worker.nr_pre - worker.nr_post) > 3:  # if post_request not called, diff increases
            worker.log.critical("[pre_request] worker %s's request not processing. kill worker." % worker.pid)
            worker.nr = worker.max_requests
            **os.kill(worker.pid, signal.SIGKILL)**  # kill the worker
            
        worker.nr_pre += 1  # increase pre_request count

    def post_request(worker, req, environ, resp):
        worker.nr_post += 1  # increase post_request count
        worker.nr_pre = worker.nr_post

        # Called after a worker processes the request.
        worker.log.debug("[post_request] %s %s - worker pid %s|ppid %s|age %s|nr %s|max_requests %s|alive %s|nr_pre %s|nr_post %s" % (
            req.method, req.path, worker.pid, worker.ppid, worker.age, worker.nr,
            worker.max_requests, worker.alive, worker.nr_pre, worker.nr_post))

**Comment by tilgovi** [comment]
That looks right, but if you ever get more than 3 requests at the same time that might kill the worker.

**Comment by tilgovi** [comment]
I am also not sure that condition will happen with the issue you describe. I will be interested to see the results!

**Comment by inyong7ee** [comment]
On server-1, read timeout issue occurred again on May 20th, which is only 11 days after issue on May 9th.

If the number of pre_request and post_request differ by more than 3, the code in pre_request hook to kill the existing worker and create a new worker behaves as expected, but after that, the newly created worker shows the same symptoms as the existing worker.
The gunicorn master worker (Arbiter) does not seem to be able to create a new worker that works normally for some reason at that time.

In the meantime, we have restarted only gunicorn and celery when restarting the service every day, but when restarting manually after an issue occurs, we restarted rabbitmq, gunicorn, celery, and flower. So we modified it to include rabbitmq and flower at restart every day.

Server-2 had an issue on May 10th.
On May 21, only 11 days after the last issue, when server-2 was restarted in the morning, rabbitmq, gunicorn, celery, and flower were all restarted, and rabbitmq's queue was also reset.
We tested sending test requests from the morning with completely new instances of the related service, and a timeout issue occurred on server-2 again.
The max_requests value for gunicorn has also been reduced from 1000 to 300, but this happens again on the 11th day, regardless of this change.

After the issue, we tried to restart only one service at a time, unlike the previous one. (Previously manually restarted rabbitmq, gunicorn, celery and flower all at once)

With the defunct process created on the system, only the gunicorn was restarted. At this time, the code for the restarting new walker was successfully executed, but the newly created walker was also defuncted.

After this test, we restarted celery. At this time gunicorn's response code worked, and the new worker was working normally.

In summary, we are currently assuming that it hangs in the delay() function for celery's shared_task which sends a message to rabbitmq on each API processing function.
However it's really strange that gunicorn never processes WORKER TIMEOUT in this case while not reaching 'finally' statement of handle_request() function in base_async.py.

**Comment by inyong7ee** [comment]
We have created a daemon program that monitors the entire process and restarts the celery master when the "[gcc] <defunct>" process occurs and kills the abnormal gunicorn workers. Although the root cause was not revealed, the issue was resolved.
